docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,452 @@
+
+- Designing a CRD like HelmChartProxy to specify an add-on to install on certain Clusters.
+- Using a `clusterSelector` to install an add-on on a selected workload Cluster and to uninstall an add-on a Cluster that is unselected.
+- Using Go templates in a field like `valuesTemplate` to dynamically resolve configurations based on each Cluster's definition.
+- Designing a CRD like HelmReleaseProxy to:
+  - Maintain an inventory of add-ons and differentiate between managed add-ons and out of band add-ons.
+  - Provide a modular design such that the HelmChartProxy controller does not need to implement any Helm operations and only make Kubernetes client calls to create, update, or delete a HelmReleaseProxy.
+
+## Cluster API enhancements
+
+Despite the fact that most of the heavy lifting for add-on management is delegated to external ClusterAddonProvider components, over time we would like to provide Cluster API’s users an integrated and seamless user experience for add-on orchestration.
+
+### clusterctl
+
+ClusterAddonProvider is a new component to be installed in a Management Cluster, and in order to make this operation simpler, we can extend clusterctl in order to manage ClusterAddonProvider implementations in clusterctl init, clusterctl upgrade, etc.
+
+Note: clusterctl assumes all the providers to be linked to a CAPI contract; while it is not clear at this stage if the same applies to ClusterAddonProvider. This should be further investigated.
+
+## Alternatives
+
+### Why not ClusterResourceSets?
+
+This proposal introduces ClusterAddonProviders as opposed to overhauling ClusterResourceSets for the following reasons:
+
+- ClusterResourceSets violate the Cluster API design philosophy of defining a desired state through a declarative spec and having controllers create, delete, or update resources to build the desired state. Instead, ClusterrResourceSets are designed to "apply once and forget."
+- Users are already using package management tools, and being able to use their preferred package management tool reduces user friction towards adopting an add-on solution.
+- ClusterResourceSets have relatively fewer user experts compared to existing package management tools like Helm.
+- ClusterResourceSets fail to take advantage of functionatlity in existing package management tools like dependency installation, and evolving ClusterResourceSets would require re-inventing the functionality of existing package management tools.
+- ClusterResourceSets do not orchestrate deletion of add-ons and was not designed with that in mind. In order to orchestrate deletion of add-ons, it would require a significant redesign of the code and interface. Doing so would still not address the above issues compared to designing a new solution with all of this in mind.
```

Comment:
```suggestion
- ClusterResourceSets do not orchestrate deletion of add-ons and was not designed with that in mind. In order to orchestrate deletion of add-ons, it would require a significant redesign of the code and interface. Doing so would still not address the above issues compared to designing a new solution with all of this in mind.


### Flux

- Another alternative is to package Flux with cluster-api. Flux is a tool for keeping Kubernetes clusters in sync with sources of configuration (like Git repositories), and automating updates to configuration when there is new code to deploy.
- Flux offers [direct integration with cluster-API ](https://fluxcd.io/flux/components/helm/helmreleases/#remote-clusters--cluster-api) by using a field built in `HelmRelease` API type. In its current implementation, it would require a `HelmRelease` per application per cluster.
```